components/tabs/src/storage.rs

```diff
@@ -146,29 +149,36 @@ impl TabsStorage {
         self.local_tabs.borrow_mut().replace(local_state);
     }
 
+    // We try our best to fit as many tabs in a payload as possible, this includes
+    // limiting the url history entries, title character count and finally drop enough tabs
+    // until we have small enough payload that the server will accept
```

Comment:
I wonder if it makes sense to call this sanitization code in `update_local_state`, especially given we hard-code the max payload size. Let's say that someone gives us 10MB worth of tab data, we will hold on to that 10MB in memory forever and every sync we will do the same trimming. OTOH, doing it in `update_local_state` means we have less in-memory and each sync doesn't need to recompute.

OTOH though, if `update_local_state` ends up being called many times before we actually sync, then it means we'd do this trimming even though we never upload it. Even on desktop, we have a 5 second time before upload, so there is a good chance this would be called a number of time before actually syncing.

So I *think* I've convinced myself the way you have done it is fine and that the extra memory overhead is probably OK when compared against the extra CPU overhead, but I'm leaving this comment in-place just incase someone else wants to argue otherwise :)